### PR TITLE
bpo-18283: Add support for bytes to shutil.which

### PR DESCRIPTION
**Pull request by csabella**
Original patch by Victor Stinner.

https://bugs.python.org/issue18283

### CODE REVIEW

File: Lib/shutil.py [hunk 1]
```diff
@@ -1279,6 +1279,13 @@ def get_terminal_size(fallback=(80, 24)):
 
     return os.terminal_size((columns, lines))
 
+# Check that a given file can be accessed with the correct mode.
+# Additionally check that `file` is not a directory, as on Windows
+# directories pass the os.access check.
+def _access_check(fn, mode):
```
Comment (tirkarthi) [inline Lib/shutil.py:1286]: Just curious any reason over moving this outside `which`?

Comment (vstinner) [reply Lib/shutil.py:1286]: As an author of the original patch, I can explain: it's inefficient to define a temporary function for each which() call. Moreover, there is no technical reason to use a nested function.

File: Lib/shutil.py [hunk 2]
```diff
@@ -1308,15 +1308,25 @@ def _access_check(fn, mode):
         path = os.environ.get("PATH", os.defpath)
     if not path:
         return None
-    path = path.split(os.pathsep)
+    if isinstance(cmd, bytes):
```
Comment (vstinner) [inline Lib/shutil.py]: Maybe add a "use_bytes = isinstance(cmd, bytes)" variable?

File: Lib/shutil.py [hunk 3]
```diff
@@ -1308,15 +1308,25 @@ def _access_check(fn, mode):
...
+        curdir = os.curdir
+        if isinstance(cmd, bytes):
+            curdir = os.fsencode(curdir)
+        if curdir not in path:
+            path.insert(0, curdir)
 
         # PATHEXT is necessary to check on Windows.
         pathext = os.environ.get("PATHEXT", "").split(os.pathsep)
+        if isinstance(cmd, bytes):
+            pathext = list(map(os.fsencode, pathext))
```
Comment (vstinner) [inline Lib/shutil.py]: Maybe use a list-comprehension instead? I'm not sure what is the best ;-)

File: Lib/test/test_shutil.py [hunk 4]
```diff
@@ -1604,6 +1607,16 @@ def test_empty_path_no_PATH(self):
...
 
 
+class TestWhichBytes(TestWhich):
+    def setUp(self):
+        TestWhich.setUp(self)
+        self.dir = os.fsencode(self.dir)
+        self.file = os.fsencode(self.file)
+        self.temp_file.name = os.fsencode(self.temp_file.name)
+        self.curdir = os.fsencode(self.curdir)
+        self.ext = b".EXE"
```
Comment (vstinner) [inline Lib/test/test_shutil.py]:
```suggestion
        self.ext = os.fsencode(self.ext)
```

**Review by vstinner — APPROVED** [review]
LGTM... well, I wrote the initial patch, so I'm not fair :-)

Just a few suggestions.

File: Lib/shutil.py (hunk 3 above)
Comment (eamanu) [reply Lib/shutil.py]: I think that ```list(map(os.fsencode, pathext))``` it's ok.

**Review by eamanu — APPROVED** [review]
LGTM. it's work correctly

File: Lib/shutil.py (hunk 3 above)
Comment (csabella) [reply Lib/shutil.py]: I did a timeit on both and the `list(map())` is slightly faster (on my system), but not much.  If it code after this didn't (potentially) iterate over `pathext` twice, a generator would be ideal.  

Also, the following `if` clause uses:
```
    if any(cmd.lower().endswith(ext.lower()) for ext in pathext):
        files = [cmd]
```

Removing the `cmd.lower()` from the `any` would be a processing optimization, but it's unrelated to this change, so I didn't put it in.
```
    cmd_lower = cmd.lower()
    if any(cmd_lower.endswith(ext.lower()) for ext in pathext):
        files = [cmd]
```

File: Lib/shutil.py (hunk 2 above)
Comment (csabella) [reply Lib/shutil.py]: Good suggestion!

File: Lib/test/test_shutil.py (hunk 4 above)
Comment (csabella) [reply Lib/test/test_shutil.py]: Thanks!

File: Lib/test/test_shutil.py [hunk 5]
```diff
@@ -1604,6 +1607,16 @@ def test_empty_path_no_PATH(self):
...
 
 
+class TestWhichBytes(TestWhich):
+    def setUp(self):
+        TestWhich.setUp(self)
+        self.dir = os.fsencode(self.dir)
+        self.file = os.fsencode(self.file)
+        self.temp_file.name = os.fsencode(self.temp_file.name)
+        self.curdir = os.fsencode(self.curdir)
+        self.ext = os.fsencode(".EXE")
```
Comment (vstinner) [inline Lib/test/test_shutil.py]: nitpick!

```suggestion
        self.ext = os.fsencode(self.ext)
```

**Review by vstinner — COMMENTED** [review]
One last remark :-)

**Comment by vstinner** [comment]
@giampaolo: I wrote the initial patch, so I would prefer that someone else would review it. Would mind to review this change please?

File: Lib/test/test_shutil.py (hunk 5 above)
Comment (csabella) [reply Lib/test/test_shutil.py]: Sorry about that.  I meant to do that on the last push.

**Comment by giampaolo** [comment]
@vstinner no problem; patch LGTM.

**Comment by vstinner** [comment]
Thanks @giampaolo for the review ;-)

**Comment by vstinner** [comment]
Thanks @csabella, I merged your PR ;-)

**Pull request merged**